nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -925,48 +886,28 @@ async fn ssc_call_pantry_attach_for_disk(
     let params = sagactx.saga_params::<Params>()?;
     let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
 
-    let pantry_address = sagactx.lookup::<SocketAddrV6>("pantry_address")?;
-
-    let endpoint = format!("http://{}", pantry_address);
-
-    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
-        .disk_id(params.disk_id)
-        .fetch_for(authz::Action::Modify)
-        .await
-        .map_err(ActionError::action_failed)?;
-
-    let disk_volume = osagactx
-        .datastore()
-        .volume_checkout(disk.volume_id)
-        .await
-        .map_err(ActionError::action_failed)?;
-
-    info!(
-        log,
-        "sending attach for disk {} volume {} to endpoint {}",
-        params.disk_id,
-        disk.volume_id,
-        endpoint,
-    );
-
-    let volume_construction_request: crucible_pantry_client::types::VolumeConstructionRequest =
-        serde_json::from_str(&disk_volume.data()).map_err(|e| {
-            ActionError::action_failed(Error::internal_error(&format!(
-                "failed to deserialize disk {} volume data: {}",
-                disk.id(),
-                e,
-            )))
-        })?;
-
-    let client = crucible_pantry_client::Client::new(&endpoint);
+    let (pantry_address, disk_already_attached_to_pantry) =
+        sagactx.lookup::<(SocketAddrV6, bool)>("pantry_address")?;
 
-    let attach_request = crucible_pantry_client::types::AttachRequest {
-        volume_construction_request,
-    };
+    if !disk_already_attached_to_pantry {
+        info!(
+            log,
+            "attaching disk {:?} to pantry at {:?}",
+            params.disk_id,
+            pantry_address
+        );
 
-    retry_until_known_result!(log, {
-        client.attach(&params.disk_id.to_string(), &attach_request)
-    });
+        call_pantry_attach_for_disk(
+            &log,
+            &opctx,
+            &osagactx.nexus(),
+            params.disk_id,
+            pantry_address,
+        )
+        .await?;
+    } else {
+        info!(log, "disk {} already attached to a pantry", params.disk_id);
```

Comment:
When would we expect a disk already attached to a pantry?

Reply:
In the case that the `snapshot_create` saga is a sub-saga to the `finalize` saga.